design/open/openapi-multipart-upload.md

```diff
@@ -0,0 +1,215 @@
+# OpenAPI for S3 Like Multipart Upload Proposal
+
+## Introduction
+
+This proposal outlines the development of Open API endpoints to enable multipart upload capabilities similar to Amazon S3 multipart. The intention is to enhance the data uploading process for large files, ensuring efficient and reliable data transfer.
+
+## Goal
+
+- Implementing multi-part upload APIs with support for parallel upload optimizes data transfer for large files over varying network conditions using lakectl.
+
+## Non-goals
+
+- Support all underlying storage types as we can experiment with the one that currently support this feature.
+- SDK to wrap the new API and support this capability.
+
+
+## Features and Functionalities
+
+In order to minimaize the scope of this feature and understand the benefits before supporting all blockstores, this feature will be scoped to S3 blockstore and only if presigned capability is enabled.
+
+Initiation of Upload: The API will allow clients to initiate a multipart upload session, assigning a unique upload ID for subsequent operations. The client will require pass the number of parts it requires and the call will provide a set of URLs to upload each part.
+
+Uploading Parts: Clients can upload individual parts of the file in parallel or in sequence. Each part will be based on the presigned URL provided by the initial call.
+
+Support for Large Files: The API will handle files of substantial sizes, ensuring that large datasets can be uploaded without issues. Minimum part size will be 5M, except for the last block.
+
+Data Integrity Checks: Each uploaded part will require the client to store the ETag as done today for presigned URL upload. The value of the ETag will be provided when the upload of all the parts is completed.
+
+Pause and Resume: Clients can pause the upload and resume later, leveraging the upload ID to reinitiate the process. (TDB need to specify how long we will keep and if cleanup will be done)
+
+Completion of Upload: Once all parts are uploaded, the client will send a request to complete the upload, address the file in lakeFS as a single file. Upload compete request will include the ETag for check part, in case of mismatch or fail to provide bad request code will return.
+
+Cancelation: In order to cancel partial upload of multipart request, the client will require to call cancellation with the upload ID provided.
+
+### Multipart upload capability using OpenAPI specification
+
+**Paths for multipart upload operations**
+
+```yaml
+  /repositories/{repository}/branches/{branch}/staging/multipart:
```

Comment:
Good point!
- Didn't want to open another interface to upload data through lakefs
- The requests to create/complete/abort multipart works on upload ID that represents the operation. The presined parts currently returned on CreateMultipartUpload operations only if you specify the number of parts you like to get and it is optional. It means that if we like in the future to open upload part data through lakeFS we can still provide this API.
- I will update the spec to rename `parts` to `presigned_parts` to indicate this number represents the number of urls the call will return.

Reply:
I don't see the alternative as uploading through lakeFS. I see the non-pre-signed version of this using local credentials on the client, i.e., the client uses the AWS SDK to upload directly to an S3, according to a physical address returned from this endpoint.  

But I'm also ok with not doing this now and seeing if there's a need there.

Reply:
Both ways - using presign and non-presigned URLs require local credentials to work.
Supporting multipart upload using the API while data transfer goes though lakeFS is possible, but I would like to avoid it and just provide more ways to enable the client direct access to the specific object using the underlying object store.